### ISSUE TITLE
inconsistent event sequences when closing a tab

### ISSUE THREAD

**Issue by andrewfaulring** [c9/core#110]
When the user closes a tab the Cloud9 IDE emits a series of events. However, if the user deletes an open file using the context menu in the Workspace panel, the Cloud9 IDE only emits a subset of the events.

First, the correct behavior. Start with a single file opened in the IDE. The file should be saved and have the focus. Now close that file by pressing the "x" in the tab header. The IDE emits the following sequence of events:

* tabAfterClose (plugins/c9.ide.editors/tabmanager.js#L1624)							
* deactivate (plugins/c9.ide.editors/tab.js#L99)
* blur (plugins/c9.ide.editors/tabmanager.js#L1547)
* tabDestroy (plugins/c9.ide.editors/tabmanager.js#L1639)

To reproduce the bug, again start with a single file opened in the IDE. The file should be saved and have the focus. Now delete the opened file using the context menu in the Workspace panel. The Cloud9 IDE now only emits

* blur (plugins/c9.ide.editors/tabmanager.js#L1547)
* tabDestroy (plugins/c9.ide.editors/tabmanager.js#L1639)

Note that both the tabAfterClose and deactivate events are not emitted.

Furthermore in this case, the result returned by calling tabManager.getState(null, true) from within a tabDestroy event listener is wrong. The result contains the destroyed tab. However, calling tabManager.findTab() with the name of the destroyed tab correctly returns undefined. For example, if the destroyed tab name is "tab0", tabManager.findTab("tab0") returns undefined. This demonstrates an internal inconsistency in the tabManager: getState() still returns the destroyed tab, whereas findTab() returns undefined.

Finally, when the user deletes an open file in the Terminal using the rm command, the Cloud9 IDE correctly emits all four events. Hence, the problem does not appear to be directly caused by deleting the open file.

**Comment by nightwing** [comment]
Should be fixed by https://github.com/c9/core/commit/b6db4698a814fc41ef7e630072ba358e5e65c66f

**Comment by andrewfaulring** [comment]
@nightwing Thanks. That update fixes the problem for the use cases that I reported.

However, I found another use case that is still broken. If you delete a folder containing an open file, the same problem occurs: two missing events. As before, if you rm -rf the folder from a Terminal, all four events are emitted.

I think the problem can be fixed by applying the same change, tab.unload() to tab.close(), to the "afterRmdir" listener

* plugins/c9.ide.editors/tabmanager.js#L148

I'm not sure if "error" listener should be change too

* plugins/c9.ide.editors/tabmanager.js#L159
